utils/flink-components-testkit/src/main/scala/pl/touk/nussknacker/engine/flink/util/test/FlinkTestScenarioRunner.scala

```diff
@@ -30,18 +30,43 @@ object testComponents {
   ).map(cd => ComponentDefinition(cd._1, cd._2))
 }
 
-case class SinkForList[T]() extends SinkForType[List[T]]
-
 class FlinkTestScenarioRunner(val components: List[ComponentDefinition], val config: Config, flinkMiniCluster: FlinkMiniClusterHolder) extends ClassBasedTestScenarioRunner {
 
   var testComponentHolder: TestComponentHolder = _
 
-  override def runWithData[I:ClassTag, R](scenario: CanonicalProcess, data: List[I]): RunnerResult[R] = {
+  override def runWithData[I: ClassTag, R](scenario: CanonicalProcess, data: List[I]): RunnerListResult[R] = {
+    implicit val typeInf: TypeInformation[I] = TypeInformation.of(implicitly[ClassTag[I]].runtimeClass.asInstanceOf[Class[I]])
+    val componentsWithData = testComponents.withDataList(data)
+    val componentsWithTestComponents = this.components ++ componentsWithData
+    run(scenario, new TestServiceInvocationCollector(testComponentHolder.runId), Some(testComponentHolder.runId), componentsWithTestComponents).map { _ =>
+      collectResults()
+    }
+  }
+
+  /**
+   * Can be used to test Flink bounded sources - we wait for the scenario to finish.
+   */
+  def runAndCollectResults[R](scenario: CanonicalProcess): RunnerListResult[R] = {
+    run(scenario, ProductionServiceInvocationCollector, None).map { _ =>
+      collectResults()
+    }
+  }
 
+  /**
+   * Can be used to test Flink based sinks.
+   */
+  def runWithDataIgnoringResults[I: ClassTag](scenario: CanonicalProcess, data: List[I]): RunnerResult[Unit] = {
     implicit val typeInf: TypeInformation[I] = TypeInformation.of(implicitly[ClassTag[I]].runtimeClass.asInstanceOf[Class[I]])
-    val modelData = LocalModelData(config, new EmptyProcessConfigCreator)
     val componentsWithData = testComponents.withDataList(data)
     val componentsWithTestComponents = this.components ++ componentsWithData
+    run(scenario, ProductionServiceInvocationCollector, None, componentsWithTestComponents)
```

Comment:
You register here `invocationCollector` component, because it is done implicitly in  `run` method. Can you refactor this code a bit, so preparing all test components (source, noop and this invocationCollector) will be on the same level? It will be easier to analyze how all those methods are different from each other.